src/parser/Parser.ts

```diff
@@ -777,24 +768,36 @@ export class Parser {
      */
     private callExpressions = [];
 
-    private functionDeclaration(isAnonymous: true, checkIdentifier?: boolean, forClassMethod?: boolean): FunctionExpression;
-    private functionDeclaration(isAnonymous: false, checkIdentifier?: boolean, forClassMethod?: boolean): FunctionStatement;
-    private functionDeclaration(isAnonymous: boolean, checkIdentifier = true, forClassMethod = false) {
+
+    private functionStatement(options: FunctionDeclarationParseOptions): FunctionStatement {
+        options.hasName = true;
+        const funcResult = this.functionDeclaration(options);
+        if (funcResult) {
+            let result = new FunctionStatement(funcResult.name, funcResult.functionExpression, this.currentNamespaceName);
+            funcResult.functionExpression.functionStatement = result;
+            if (!options.onlyCallableAsMember) {
+                this._references.functionStatements.push(result);
+            }
+            return result;
+        }
+    }
+
+    private functionDeclaration(options: FunctionDeclarationParseOptions = {}): { name: Identifier; functionExpression: FunctionExpression } {
```

Comment:
I don't particularly like that I have to return an object wrapper here, but I didn't know how else to do it without duplicating the function name finding code.